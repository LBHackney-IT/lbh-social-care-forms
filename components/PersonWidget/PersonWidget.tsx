import React from 'react';
import { LegacyResident, Resident } from 'types';
import s from './PersonWidget.module.scss';
import { format } from 'date-fns';

const prettyDate = (isoDateString: string): string => {
  const parsed = new Date(isoDateString);
  try {
    return format(parsed, 'd MMM yyyy');
  } catch (e) {
    return '';
  }
};

interface Props {
  person: Resident | LegacyResident;
  grouped?: boolean;
  index?: number;
  onRemove?: (value: number) => void;
  open?: boolean;
  setOpen?: (value: number | false) => void;
}

const PersonWidget = ({
  person,
  grouped,
  onRemove,
  index,
  open,
  setOpen,
}: Props): React.ReactElement => {
  const dateOfBirth = prettyDate(person?.dateOfBirth ?? '');
  const displayAddress = person?.address;
  const firstDisplayAddressMatch =
    'id' in person
      ? person?.addresses?.find((address) => address.isDisplayAddress === 'Y')
      : person.address;

  if (grouped)
    return (
      <aside className={s.aside}>
        <details className={s.details} open={open}>
          <summary
            className={s.summary}
            onClick={(e) => {
              e.preventDefault();
              setOpen &&
                (open ? setOpen(false) : index !== undefined && setOpen(index));
            }}
          >
            <h2 className={`lbh-heading-h3 ${s.title}`}>
              {person.firstName} {person.lastName}
            </h2>

            <svg width="17" height="10" viewBox="0 0 17 10">
              <path d="M2 1.5L8.5 7.5L15 1.5" strokeWidth="3" />
            </svg>
          </summary>

          {dateOfBirth && (
            <p className={`lbh-body-s ${s.paragraph}`}>Born {dateOfBirth}</p>
          )}
          {displayAddress && (
            <p
              className={`lbh-body-s ${s.paragraph}`}
              data-testid="resident-display-address"
            >
              {displayAddress?.address}
              <br />
              {displayAddress?.postcode}
            </p>
          )}
          {firstDisplayAddressMatch && (
            <p
              className={`lbh-body-s ${s.paragraph}`}
              data-testid="resident-last-address"
            >
              {'address' in firstDisplayAddressMatch
                ? firstDisplayAddressMatch.address
                : firstDisplayAddressMatch.addressLines}
              <br />
              {'postcode' in firstDisplayAddressMatch
                ? firstDisplayAddressMatch.postcode
                : firstDisplayAddressMatch.postCode}
            </p>
          )}

          <button
            className="lbh-link lbh-body-s"
            onClick={() =>
              onRemove &&
              onRemove(
                'id' in person ? person.id : parseInt(String(person.mosaicId))
              )
            }
          >
            Remove
          </button>
        </details>
      </aside>
    );

  return (
    <aside className={s.aside}>
      <h2 className={`lbh-heading-h3 ${s.title}`}>
        {person.firstName} {person.lastName}
      </h2>

      {dateOfBirth && (
        <p className={`lbh-body-s ${s.paragraph}`}>Born {dateOfBirth}</p>
      )}
      {displayAddress && (
        <p
          className={`lbh-body-s ${s.paragraph}`}
          data-testid="resident-display-address"
        >
          {displayAddress?.address}
          <br />
          {displayAddress?.postcode}
        </p>
      )}
      {firstDisplayAddressMatch && (
        <p
          className={`lbh-body-s ${s.paragraph}`}
          data-testid="resident-last-address"
        >
          {'address' in firstDisplayAddressMatch
            ? firstDisplayAddressMatch.address
            : firstDisplayAddressMatch.addressLines}
          <br />
          {'postcode' in firstDisplayAddressMatch
            ? firstDisplayAddressMatch.postcode
            : firstDisplayAddressMatch.postCode}
        </p>
      )}
    </aside>
  );
};

export default PersonWidget;
