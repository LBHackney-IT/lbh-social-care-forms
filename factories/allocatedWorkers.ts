import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
