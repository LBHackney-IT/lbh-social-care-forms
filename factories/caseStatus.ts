import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
