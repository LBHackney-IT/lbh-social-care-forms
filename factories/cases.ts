import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
