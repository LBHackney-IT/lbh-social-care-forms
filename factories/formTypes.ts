import { Factory } from 'fishery';
import { FormValue, FormStatusValue } from 'types';

export const mockedFormValueFactory = Factory.define<FormValue>(() => ({
  name: 'test',
  description: 'description',
  options: [mockedFormStatusValueFactory.build()],
}));

export const mockedFormStatusValueFactory = Factory.define<FormStatusValue>(
  () => ({
    name: 'N1',
    description: 'Noo 1',
  })
);

export const mockedFormValue = [mockedFormValueFactory.build()];
