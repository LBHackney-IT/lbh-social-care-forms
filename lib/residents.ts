import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, User } from 'types';
import { getAuditingParams } from '../utils/auditing';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends LegacyResident {
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(({ addressList, ...resident }: ResidentBE): LegacyResident => {
    const address = addressList?.find(
      ({ displayAddressFlag }) => displayAddressFlag === 'Y'
    );
    return {
      ...resident,
      address: address && {
        address: address.addressLine1,
        postcode: address.postCode,
        uprn: address.uprn,
      },
    };
  });

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  user: User
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params: getAuditingParams(user),
  });
  return data;
};

export const normalisePhoneInput = (
  formData: Resident
): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Resident
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (formData: Resident): Promise<Resident> => {
  await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return formData;
};
