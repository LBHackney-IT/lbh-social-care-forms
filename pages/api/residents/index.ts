import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';
import { apiHandler } from 'lib/apiHandler';
import { middleware as csrfMiddleware } from 'lib/csrfToken';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    res.status(StatusCodes.UNAUTHORIZED);
    return;
  }
  if (!user.isAuthorised) {
    res.status(StatusCodes.FORBIDDEN);
    return;
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Residents get error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Resident post error:',
          (error as AxiosError)?.response?.data
        );
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
