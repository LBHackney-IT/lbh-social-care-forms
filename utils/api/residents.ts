import axios from 'axios';
import useSWR, { useSWRInfinite, SWRResponse, SWRInfiniteResponse } from 'swr';

import { getInfiniteKey } from 'utils/api';

import type { Resident, ErrorAPI } from 'types';

export const useResidents = (
  params: Record<string, unknown>,
  invoke = true
): SWRInfiniteResponse<Record<string, unknown>, Error> =>
  useSWRInfinite(
    // @ts-ignore
    invoke ? getInfiniteKey('/api/residents', 'residents', params) : null
  );

export const useResident = (
  personId: number
): SWRResponse<Resident, ErrorAPI> => useSWR(`/api/residents/${personId}`);

export const addResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/residents`, formData);
  return { ref: data?.id, data };
};

export const updateResident = async (
  personId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents/${personId}`, formData);
  return { ref: data?.id, data };
};

export const patchResident = async (formData: {
  id: number;
  createdBy: string;
  [key: string]: unknown;
}): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents`, formData);
  return { ref: data?.id, data };
};
