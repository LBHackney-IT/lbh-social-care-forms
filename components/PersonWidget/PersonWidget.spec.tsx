import PersonWidget from './PersonWidget';
import { render, fireEvent } from '@testing-library/react';
import { residentFactory } from 'factories/residents';

describe('PersonWidget', () => {
  it('should render properly', () => {
    const resident = residentFactory.build();
    const { asFragment } = render(<PersonWidget person={resident} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when there is a person', () => {
    const resident = residentFactory.build();

    const { getByText, queryByText } = render(
      <PersonWidget person={resident} />
    );

    expect(getByText('Foo Bar'));
    expect(getByText('Born 13 Nov 2020'));
    expect(queryByText('FLAT 10, GEORGE LEYBOURNE HOUSE, FLETCHER STREET'));
    expect(queryByText('E1 8HW'));
  });

  it('should handle when there is no date of birth', () => {
    const resident = residentFactory.build();
    resident.dateOfBirth = '';

    const { queryByText } = render(<PersonWidget person={resident} />);

    expect(queryByText('Born')).toBeNull();
  });

  it('should handle when there is an invalid date of birth', () => {
    const resident = residentFactory.build();
    resident.dateOfBirth = 'this is not a date';

    const { getByText, queryByText } = render(
      <PersonWidget person={resident} />
    );

    expect(getByText('Foo Bar'));
    expect(queryByText('Born')).toBeNull();
  });

  it('should handle when there is no address', () => {
    const resident = residentFactory.build();
    resident.addresses = [];

    const { queryByText } = render(<PersonWidget person={resident} />);

    expect(
      queryByText('FLAT 10, GEORGE LEYBOURNE HOUSE, FLETCHER STREET')
    ).toBeNull();
  });

  it('should return the last address if there are multiple addresses saved on the person', () => {
    const firstAddress = {
      addressLines: '100 Example Street',
      postCode: 'E1 8HW',
    };

    const newAddress = {
      addressLines: '123 TestCode Lane',
      postCode: 'SE1 7DT',
    };

    const resident = residentFactory.build();
    resident.addresses = [firstAddress, newAddress];

    const { getByTestId, asFragment } = render(
      <PersonWidget person={resident} />
    );

    expect(getByTestId('resident-last-address')).toHaveTextContent(
      '123 TestCode Lane'
    );
    expect(getByTestId('resident-last-address')).toHaveTextContent('SE1 7DT');

    expect(getByTestId('resident-last-address')).not.toHaveTextContent(
      '100 Example Street'
    );

    expect(getByTestId('resident-last-address')).not.toHaveTextContent(
      'E1 8HW'
    );
  });

  it('should call the setOpen callback with the id of the selected resident if the widget is not open', () => {
    const resident = residentFactory.build();
    const mockSetOpen = jest.fn();

    const index = 1;

    const { container } = render(
      <PersonWidget
        person={resident}
        grouped={true}
        onRemove={jest.fn()}
        index={index}
        open={false}
        setOpen={mockSetOpen}
      />
    );

    const input = container.querySelector('summary');
    input && fireEvent.click(input);

    expect(mockSetOpen).toBeCalledWith(index);
  });

  it('should call the setOpen callback with false when closing a widget', () => {
    const resident = residentFactory.build();
    const mockSetOpen = jest.fn();

    const { container } = render(
      <PersonWidget
        person={resident}
        grouped={true}
        onRemove={jest.fn()}
        index={1}
        open={true}
        setOpen={mockSetOpen}
      />
    );

    const input = container.querySelector('summary');
    input && fireEvent.click(input);

    expect(mockSetOpen).toBeCalledWith(false);
  });

  it('should call the onRemove callback with the id of the resident', () => {
    const resident = residentFactory.build();
    const mockRemove = jest.fn();

    const index = 1;

    const { queryByText } = render(
      <PersonWidget
        person={resident}
        grouped={true}
        onRemove={mockRemove}
        index={index}
        open={false}
        setOpen={jest.fn()}
      />
    );

    const removeButton = queryByText('Remove');
    removeButton && fireEvent.click(removeButton);

    expect(mockRemove).toBeCalledWith(resident.id);
  });

  // calling onRemove
});
