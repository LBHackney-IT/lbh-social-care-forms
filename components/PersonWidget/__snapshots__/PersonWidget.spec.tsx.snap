// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonWidget should render properly 1`] = `
<DocumentFragment>
  <aside
    class="aside"
  >
    <h2
      class="lbh-heading-h3 title"
    >
      Foo Bar
    </h2>
    <p
      class="lbh-body-s paragraph"
    >
      Born 13 Nov 2020
    </p>
    <p
      class="lbh-body-s paragraph"
      data-testid="resident-display-address"
    >
      sjakdjlk
      <br />
      hdsadjk
    </p>
  </aside>
</DocumentFragment>
`;
